sdks/python/apache_beam/runners/portability/expansion_service_test.py

```diff
@@ -73,39 +84,139 @@ def from_runner_api_parameter(payload, unused_context):
     return FilterLessThanTransform(payload.decode('utf8'))
 
 
-@ptransform.PTransform.register_urn('simple', None)
-class SimpleTransform(ptransform.PTransform):
+@ptransform.PTransform.register_urn(TEST_PREFIX_URN, None)
+@beam.typehints.with_output_types(unicode)
+class PrefixTransform(ptransform.PTransform):
+  def __init__(self, payload):
+    self._payload = payload
+
   def expand(self, pcoll):
-    return pcoll | 'TestLabel' >> beam.Map(lambda x: 'Simple(%s)' % x)
+    return pcoll | 'TestLabel' >> beam.Map(
+        lambda x: '{}{}'.format(self._payload, x))
 
   def to_runner_api_parameter(self, unused_context):
-    return 'simple', None
+    return TEST_PREFIX_URN, None
 
   @staticmethod
-  def from_runner_api_parameter(unused_parameter, unused_context):
-    return SimpleTransform()
+  def from_runner_api_parameter(payload, unused_context):
+    return PrefixTransform(payload.decode('utf8'))
 
 
-@ptransform.PTransform.register_urn('multi', None)
+@ptransform.PTransform.register_urn(TEST_MULTI_URN, None)
 class MutltiTransform(ptransform.PTransform):
   def expand(self, pcolls):
     return {
         'main':
             (pcolls['main1'], pcolls['main2'])
             | beam.Flatten()
             | beam.Map(lambda x, s: x + s,
-                       beam.pvalue.AsSingleton(pcolls['side'])),
-        'side': pcolls['side'] | beam.Map(lambda x: x + x),
+                       beam.pvalue.AsSingleton(pcolls['side'])
+                      ).with_output_types(unicode),
+        'side': pcolls['side'] | beam.Map(
+            lambda x: x + x).with_output_types(unicode),
     }
 
   def to_runner_api_parameter(self, unused_context):
-    return 'multi', None
+    return TEST_MULTI_URN, None
 
   @staticmethod
   def from_runner_api_parameter(unused_parameter, unused_context):
     return MutltiTransform()
 
 
+@ptransform.PTransform.register_urn(TEST_GBK_URN, None)
+class GBKTransform(ptransform.PTransform):
+  def expand(self, pcoll):
+    return pcoll | 'TestLabel' >> beam.GroupByKey()
+
+  def to_runner_api_parameter(self, unused_context):
+    return TEST_GBK_URN, None
+
+  @staticmethod
+  def from_runner_api_parameter(unused_parameter, unused_context):
```

Comment:
Do we need to implement from_runner_api_parameter() and to_runner_api_parameter() methods here and below ? don't the default implementation work ?

Reply:
Default implementation won't work in most cases. It's more like a placeholder not supposed to handle any generic uses.

https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/ptransform.py#L704